src/main/scala/inox/solvers/z3/Z3Native.scala

```diff
@@ -727,17 +718,25 @@ trait Z3Native extends ADTManagers with Interruptible { self: AbstractSolver =>
     val ex = new ModelExtractor(model)
 
     val vars = variables.aToB.flatMap {
+      /** WARNING this code is very similar to Z3Unrolling.modelEval!!! */
       case (v,z3ID) => (v.tpe match {
         case BooleanType =>
           model.evalAs[Boolean](z3ID).map(BooleanLiteral)
 
         case Int32Type =>
-          model.evalAs[Int](z3ID).map(IntLiteral(_)).orElse {
+          model.evalAs[Int](z3ID).map(Int32Literal(_)).orElse {
             model.eval(z3ID).flatMap(t => ex.get(t, Int32Type))
           }
 
-        case IntegerType =>
-          model.evalAs[Int](z3ID).map(i => IntegerLiteral(BigInt(i)))
+         /*
+          * FIXME this seems to work, but why not rely on the default evaluation instead of the specialised evaluation
+          *       for Int(32) which fails for big integers.
+          *
+          * case IntegerType =>
+          *  model.evalAs[Int](z3ID).map(IntegerLiteral(_)).orElse {
+          *    model.eval(z3ID).flatMap(ex.get(_, IntegerType))
+          *  }
+          */
```

Comment:
Feedback on this would be welcome: the removed code is buggy, the one commented out works but I'm not sure we want to use that trick (i.e. it'd be better to remove it altogether).